code/parachain/runtime/picasso/src/lib.rs

```diff
@@ -178,6 +196,78 @@ parameter_types! {
 	pub const SS58Prefix: u8 = 49;
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, Default)]
+pub struct Router {}
+
+impl ModuleRouter for Router {
+	fn get_route_mut(&mut self, _module_id: &ModuleId) -> Option<&mut dyn Module> {
+		None
+	}
+
+	fn has_route(_module_id: &ModuleId) -> bool {
+		false
+	}
+
+	fn lookup_module_by_port(_port_id: &PortId) -> Option<ModuleId> {
+		None
+	}
+}
+
+pub struct IbcDenomToAssetIdConversion;
+
+impl DenomToAssetId<Runtime> for IbcDenomToAssetIdConversion {
+	type Error = DispatchError;
+
+	fn from_denom_to_asset_id(denom: &String) -> Result<CurrencyId, Self::Error> {
+		let denom_bytes = denom.as_bytes().to_vec();
+		if let Some(id) = IbcDenoms::<Runtime>::get(&denom_bytes) {
+			return Ok(id)
+		}
+
+		let asset_id =
+			<currency_factory::Pallet<Runtime> as CurrencyFactoryT>::create(RangeId::IBC_ASSETS)?;
+
+		IbcDenoms::<Runtime>::insert(denom_bytes.clone(), asset_id);
+		IbcAssetIds::<Runtime>::insert(asset_id, denom_bytes);
+
+		let location = XcmAssetLocation::new(MultiLocation::new(
+			1,
+			X1(Junction::GeneralIndex(asset_id.into())),
+		));
+		assets_registry::Pallet::<Runtime>::set_reserve_location(
+			asset_id,
+			location,
+			Rational64::one(),
+			Some(12),
+		)?;
+
+		Ok(asset_id)
+	}
+
+	fn from_asset_id_to_denom(id: CurrencyId) -> Option<String> {
+		IbcAssetIds::<Runtime>::get(id).and_then(|denom| String::from_utf8(denom).ok())
```

Comment:
If we have an invalid denom in storage, this will silently say it is simply missing. Can we change the signature to either Option<Result<String>>, or add a log line so we know it fails because bad encoding slipped in?